Public/Src/Engine/Processes/ReportedFileAccess.cs

```diff
@@ -622,6 +623,69 @@ public bool TryParseAbsolutePath(PipExecutionContext context, Process pip, out A
                 enumeratePattern);
         }
 
+        /// <nodoc />
+        public void Serialize(
+            BuildXLWriter writer,
+            Dictionary<ReportedProcess, int> processMap,
+            Action<BuildXLWriter, AbsolutePath> writePath)
+        {
+            writer.Write((byte)Operation);
+
+            if (processMap != null && processMap.TryGetValue(Process, out int index))
+            {
+                writer.WriteCompact(index);
+            }
+            else
+            {
+                Process.Serialize(writer);
+            }
+
+            writer.WriteCompact((int)RequestedAccess);
+            writer.WriteCompact((int)Status);
+            writer.Write(ExplicitlyReported);
+            writer.Write(Error);
+            writer.Write(Usn.Value);
+            writer.Write((uint)DesiredAccess);
+            writer.Write((uint)ShareMode);
+            writer.Write((uint)CreationDisposition);
+            writer.Write((uint)FlagsAndAttributes);
+
+            if (writePath != null)
+            {
+                writePath.Invoke(writer, ManifestPath);
+            }
+            else
+            {
+                writer.Write(ManifestPath);
+            }
+
+            writer.WriteNullableString(Path);
+            writer.WriteNullableString(EnumeratePattern);
+        }
+
+        /// <nodoc />
+        public static ReportedFileAccess Deserialize(
+            BuildXLReader reader, 
+            IReadOnlyList<ReportedProcess> processes, 
+            Func<BuildXLReader, AbsolutePath> readPath)
+        {
+            return new ReportedFileAccess(
+                operation: (ReportedFileOperation)reader.ReadByte(),
+                process: processes != null ? processes[reader.ReadInt32Compact()] : ReportedProcess.Deserialize(reader),

```

Comment:
>ReportedProcess [](start = 84, length = 15)

Why the asymetry between `Process.Serialize` and `ReportedProcess.Deserialize`?

Reply:
I think that the methods look asymmetric because of additional arguments, `processMap` and `processes`. These two additional arguments are used for compacting the serialization, i.e., we don't need to repeatedly serialize the same `ReportedProcess`.

---
In reply to: [276816570](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276816570) [](ancestors = 276816570)